File: familyvault/templates.py
"""Schema definitions for each markdown type in the vault.

Each Template describes one type (Person, Place, Memory). Field order is the
order shown in the UI: most important first, least important last. A field's
``kind`` decides how the value is rendered into YAML frontmatter.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import Any


class Kind(str, Enum):
    """How a field's value is stored and edited."""

    TEXT = "text"          # single-line string
    LONGTEXT = "longtext"  # multi-line string (textarea in UI)
    DATE = "date"          # ISO date string YYYY-MM-DD
    URL = "url"            # validated as a URL
    LINK = "link"          # single wikilink to another vault file
    LINKS = "links"        # list of wikilinks
    NUMBER = "number"


@dataclass(frozen=True)
class Field:
    key: str          # YAML key in frontmatter
    label: str        # Human label shown in UI
    kind: Kind
    target_type: str | None = None  # for LINK/LINKS: which Template type
    hint: str | None = None         # short usage hint shown next to the input

    def display_hint(self) -> str:
        """Return the hint shown under the input. Falls back to a sensible
        default for the field's Kind when no explicit hint was set."""
        if self.hint is not None:
            return self.hint
        return _DEFAULT_HINTS.get(self.kind, "")


@dataclass(frozen=True)
class Template:
    type_name: str     # e.g. "Person"
    folder: str        # vault subfolder, e.g. "People"
    fields: tuple[Field, ...]

    def empty_frontmatter(self) -> dict[str, Any]:
        """Build a frontmatter dict with every field present but blank."""
        out: dict[str, Any] = {"type": self.type_name.lower()}
        for f in self.fields:
            out[f.key] = _blank(f.kind)
        return out

    def field(self, key: str) -> Field | None:
        for f in self.fields:
            if f.key == key:
                return f
        return None


def _blank(kind: Kind) -> Any:
    if kind == Kind.LINKS:
        return []
    return ""


_DEFAULT_HINTS: dict[Kind, str] = {
    Kind.LINK: "One name",
    Kind.LINKS: "Separate multiple names with commas",
    Kind.DATE: "YYYY-MM-DD",
    Kind.URL: "Paste a full URL",
    Kind.NUMBER: "A number",
}


PERSON = Template(
    type_name="Person",
    folder="People",
    fields=(
        Field("name", "Name", Kind.TEXT,
              hint="First and last, e.g. Mary Smith"),
        Field("lives_in", "Lives in", Kind.LINK, target_type="Place",
              hint="A city, town, or specific place"),
        Field("birthdate", "Birthdate", Kind.DATE),
        Field("birthplace", "Birthplace", Kind.LINK, target_type="Place"),
        Field("parents", "Parents", Kind.LINKS, target_type="Person",
              hint="Two names, separated by a comma"),
        Field("siblings", "Siblings", Kind.LINKS, target_type="Person",
              hint="Separate multiple names with commas"),
        Field("partner", "Partner", Kind.LINK, target_type="Person"),
        Field("children", "Children", Kind.LINKS, target_type="Person",
              hint="Separate multiple names with commas"),
        Field("core_memories", "Core Memories", Kind.LINKS, target_type="Memory",
              hint="Reference existing memories by title"),
        Field("paternal_grandparents", "Paternal Grandparents", Kind.LINKS, target_type="Person",
              hint="Two names, separated by a comma"),
        Field("maternal_grandparents", "Maternal Grandparents", Kind.LINKS, target_type="Person",
              hint="Two names, separated by a comma"),
        Field("extended_family", "Extended Family", Kind.LINKS, target_type="Person",
              hint="Aunts, uncles, cousins — comma-separated"),
        Field("occupation", "Occupation", Kind.TEXT),
        Field("degrees", "Degrees", Kind.LONGTEXT,
              hint="One per line, e.g. BA English — Boston U — 2004"),
        Field("past_addresses", "Past Addresses", Kind.LINKS, target_type="Place",
              hint="Cities or places they've lived, comma-separated"),
        Field("friends", "Friends", Kind.LINKS, target_type="Person",
              hint="Separate multiple names with commas"),
        Field("employers", "Employers", Kind.LONGTEXT,
              hint="One per line, e.g. Acme Corp — 2010 to 2015"),
    ),
)


PLACE = Template(
    type_name="Place",
    folder="Places",
    fields=(
        Field("name", "Name", Kind.TEXT,
              hint="What you'd call this place in conversation"),
        Field("maps_link", "Google Maps link", Kind.URL,
              hint="Paste a Google Maps URL — coordinates will be extracted"),
        Field("latitude", "Latitude", Kind.NUMBER,
              hint="Decimal degrees, e.g. 42.3601"),
        Field("longitude", "Longitude", Kind.NUMBER,
              hint="Decimal degrees, e.g. -71.0589"),
        Field("address", "Address", Kind.TEXT),
        Field("notes", "Notes", Kind.LONGTEXT,
              hint="Anything else worth remembering about this place"),
    ),
)


MEMORY = Template(
    type_name="Memory",
    folder="Memories",
    fields=(
        Field("title", "Title (what)", Kind.TEXT,
              hint="A short title for this memory"),
        Field("when", "When", Kind.DATE),
        Field("who", "Who", Kind.LINKS, target_type="Person",
              hint="People present — separate multiple names with commas"),
        Field("where", "Where", Kind.LINKS, target_type="Place",
              hint="Places involved — separate multiple with commas"),
        Field("why", "Why", Kind.LONGTEXT,
              hint="Why does this moment matter?"),
        Field("photo", "Photo", Kind.TEXT),  # filename only, embedded in body
        Field("story", "Story", Kind.LONGTEXT,
              hint="Tell it in your own words"),
    ),
)


ALL: tuple[Template, ...] = (PERSON, PLACE, MEMORY)


def by_type(type_name: str) -> Template:
    for t in ALL:
        if t.type_name.lower() == type_name.lower():
            return t
    raise KeyError(f"Unknown template type: {type_name}")

File: src/lib/templates.ts
/**
 * Schemas for each type of markdown record in the vault.
 *
 * Each Template lists its fields in display importance order (most important
 * first, least important last). Field.kind controls how the value is rendered
 * in the UI and stored in YAML frontmatter.
 *
 * Same shape as the Python `templates.py` from the Streamlit version so a
 * vault populated by either app reads identically.
 */

export type Kind =
  | 'text'      // single-line string
  | 'longtext'  // multi-line string
  | 'date'      // ISO date YYYY-MM-DD
  | 'url'       // a URL string
  | 'link'      // single [[wikilink]]
  | 'links'     // list of [[wikilinks]]
  | 'number';   // numeric

export interface Field {
  key: string;            // YAML key
  label: string;          // displayed label
  kind: Kind;
  targetType?: TypeName;  // for link/links: which Template type
  hint?: string;          // optional explicit hint; falls back to kind default
}

export type TypeName = 'Person' | 'Place' | 'Memory';

export interface Template {
  type: TypeName;
  folder: string;         // vault subfolder, e.g. "People"
  fields: readonly Field[];
}

const DEFAULT_HINTS: Record<Kind, string> = {
  text: '',
  longtext: '',
  date: 'YYYY-MM-DD',
  url: 'Paste a full URL',
  link: 'One name',
  links: 'Separate multiple names with commas',
  number: 'A number',
};

export function hintFor(field: Field): string {
  return field.hint ?? DEFAULT_HINTS[field.kind];
}

export const PERSON: Template = {
  type: 'Person',
  folder: 'People',
  fields: [
    { key: 'name', label: 'Name', kind: 'text', hint: 'First and last, e.g. Mary Smith' },
    { key: 'lives_in', label: 'Lives in', kind: 'link', targetType: 'Place', hint: 'A city, town, or specific place' },
    { key: 'birthdate', label: 'Birthdate', kind: 'date' },
    { key: 'birthplace', label: 'Birthplace', kind: 'link', targetType: 'Place' },
    { key: 'parents', label: 'Parents', kind: 'links', targetType: 'Person', hint: 'Two names, separated by a comma' },
    { key: 'siblings', label: 'Siblings', kind: 'links', targetType: 'Person' },
    { key: 'partner', label: 'Partner', kind: 'link', targetType: 'Person' },
    { key: 'children', label: 'Children', kind: 'links', targetType: 'Person' },
    { key: 'core_memories', label: 'Core Memories', kind: 'links', targetType: 'Memory', hint: 'Reference existing memories by title' },
    { key: 'paternal_grandparents', label: 'Paternal Grandparents', kind: 'links', targetType: 'Person', hint: 'Two names, separated by a comma' },
    { key: 'maternal_grandparents', label: 'Maternal Grandparents', kind: 'links', targetType: 'Person', hint: 'Two names, separated by a comma' },
    { key: 'extended_family', label: 'Extended Family', kind: 'links', targetType: 'Person', hint: 'Aunts, uncles, cousins — comma-separated' },
    { key: 'occupation', label: 'Occupation', kind: 'text' },
    { key: 'degrees', label: 'Degrees', kind: 'longtext', hint: 'One per line, e.g. BA English — Boston U — 2004' },
    { key: 'past_addresses', label: 'Past Addresses', kind: 'links', targetType: 'Place', hint: 'Cities or places they\'ve lived' },
    { key: 'friends', label: 'Friends', kind: 'links', targetType: 'Person' },
    { key: 'employers', label: 'Employers', kind: 'longtext', hint: 'One per line, e.g. Acme Corp — 2010 to 2015' },
  ],
};

export const PLACE: Template = {
  type: 'Place',
  folder: 'Places',
  fields: [
    { key: 'name', label: 'Name', kind: 'text', hint: 'What you\'d call this place in conversation' },
    { key: 'maps_link', label: 'Google Maps link', kind: 'url', hint: 'Paste a Google Maps URL' },
    { key: 'latitude', label: 'Latitude', kind: 'number', hint: 'Decimal degrees, e.g. 42.3601' },
    { key: 'longitude', label: 'Longitude', kind: 'number', hint: 'Decimal degrees, e.g. -71.0589' },
    { key: 'address', label: 'Address', kind: 'text' },
    { key: 'notes', label: 'Notes', kind: 'longtext' },
  ],
};

export const MEMORY: Template = {
  type: 'Memory',
  folder: 'Memories',
  fields: [
    { key: 'title', label: 'Title (what)', kind: 'text', hint: 'A short title for this memory' },
    { key: 'when', label: 'When', kind: 'date' },
    { key: 'who', label: 'Who', kind: 'links', targetType: 'Person', hint: 'People present — separate multiple with commas' },
    { key: 'where', label: 'Where', kind: 'links', targetType: 'Place', hint: 'Places involved — separate multiple with commas' },
    { key: 'why', label: 'Why', kind: 'longtext', hint: 'Why does this moment matter?' },
    { key: 'photo', label: 'Photo', kind: 'text' },
    { key: 'story', label: 'Story', kind: 'longtext', hint: 'Tell it in your own words' },
  ],
};

export const TEMPLATES: readonly Template[] = [PERSON, PLACE, MEMORY];

export function byType(type: TypeName | string): Template {
  const lower = type.toLowerCase();
  const found = TEMPLATES.find((t) => t.type.toLowerCase() === lower);
  if (!found) throw new Error(`Unknown template type: ${type}`);
  return found;
}

export function blankValue(kind: Kind): unknown {
  return kind === 'links' ? [] : '';
}

export function emptyFrontmatter(template: Template): Record<string, unknown> {
  const out: Record<string, unknown> = { type: template.type.toLowerCase() };
  for (const f of template.fields) out[f.key] = blankValue(f.kind);
  return out;
}

export function findField(template: Template, key: string): Field | undefined {
  return template.fields.find((f) => f.key === key);
}
